File: tests/fixtures/go-gin/service/email_service.go
// service/email_service.go - Email service

package service

import (
	"fmt"
	"html/template"
	"strings"

	"myapp/config"
	"myapp/models"
)

type EmailService struct {
	config   *config.Config
	fromEmail string
	siteName  string
	baseURL   string
}

func NewEmailService(cfg *config.Config) *EmailService {
	return &EmailService{
		config:    cfg,
		fromEmail: cfg.Email.FromAddress,
		siteName:  cfg.App.Name,
		baseURL:   cfg.App.BaseURL,
	}
}

func (s *EmailService) SendVerificationEmail(user *models.User) error {
	token := s.generateVerificationToken(user)
	verificationURL := fmt.Sprintf("%s/verify-email?token=%s", s.baseURL, token)

	subject := fmt.Sprintf("Verify your email for %s", s.siteName)
	body := s.renderTemplate("verify_email", map[string]interface{}{
		"User":           user,
		"VerificationURL": verificationURL,
		"SiteName":       s.siteName,
	})

	return s.sendEmail(user.Email, subject, body)
}

func (s *EmailService) SendPasswordResetEmail(user *models.User, token string) error {
	resetURL := fmt.Sprintf("%s/reset-password?token=%s", s.baseURL, token)

	subject := fmt.Sprintf("Reset your password for %s", s.siteName)
	body := s.renderTemplate("password_reset", map[string]interface{}{
		"User":      user,
		"ResetURL":  resetURL,
		"SiteName":  s.siteName,
		"ExpiresIn": "24 hours",
	})

	return s.sendEmail(user.Email, subject, body)
}

func (s *EmailService) SendWelcomeEmail(user *models.User) error {
	subject := fmt.Sprintf("Welcome to %s!", s.siteName)
	body := s.renderTemplate("welcome", map[string]interface{}{
		"User":     user,
		"SiteName": s.siteName,
		"LoginURL": fmt.Sprintf("%s/login", s.baseURL),
	})

	return s.sendEmail(user.Email, subject, body)
}

func (s *EmailService) SendSuspensionNotice(user *models.User, reason string, durationDays int) error {
	subject := fmt.Sprintf("Your %s account has been suspended", s.siteName)
	body := s.renderTemplate("suspension_notice", map[string]interface{}{
		"User":         user,
		"Reason":       reason,
		"DurationDays": durationDays,
		"SiteName":     s.siteName,
		"SupportEmail": s.config.Email.SupportAddress,
	})

	return s.sendEmail(user.Email, subject, body)
}

// Q: What's the best approach for handling email delivery failures and retries in Go?
func (s *EmailService) SendBulkEmail(users []*models.User, subject string, templateName string, extraData map[string]interface{}) (*BulkEmailResult, error) {
	result := &BulkEmailResult{
		Sent:   0,
		Failed: 0,
		Errors: make([]BulkEmailError, 0),
	}

	for _, user := range users {
		data := map[string]interface{}{
			"User":     user,
			"SiteName": s.siteName,
		}
		for k, v := range extraData {
			data[k] = v
		}

		body := s.renderTemplate(templateName, data)
		err := s.sendEmail(user.Email, subject, body)

		if err != nil {
			result.Failed++
			result.Errors = append(result.Errors, BulkEmailError{
				Email: user.Email,
				Error: err.Error(),
			})
		} else {
			result.Sent++
		}
	}

	return result, nil
}

func (s *EmailService) SendNotificationEmail(user *models.User, notificationType string, data map[string]interface{}) error {
	templates := map[string]string{
		"new_login":        "new_login",
		"password_changed": "password_changed",
		"profile_updated":  "profile_updated",
		"security_alert":   "security_alert",
	}

	templateName, ok := templates[notificationType]
	if !ok {
		return fmt.Errorf("unknown notification type: %s", notificationType)
	}

	subjects := map[string]string{
		"new_login":        fmt.Sprintf("New login to your %s account", s.siteName),
		"password_changed": fmt.Sprintf("Your %s password was changed", s.siteName),
		"profile_updated":  fmt.Sprintf("Your %s profile was updated", s.siteName),
		"security_alert":   fmt.Sprintf("Security alert for your %s account", s.siteName),
	}

	templateData := map[string]interface{}{
		"User":     user,
		"SiteName": s.siteName,
	}
	for k, v := range data {
		templateData[k] = v
	}

	body := s.renderTemplate(templateName, templateData)
	return s.sendEmail(user.Email, subjects[notificationType], body)
}

func (s *EmailService) sendEmail(to, subject, body string) error {
	// Would implement actual email sending via SMTP or email service
	// For now, just log
	fmt.Printf("Sending email to %s: %s\n", to, subject)
	return nil
}

func (s *EmailService) renderTemplate(name string, data map[string]interface{}) string {
	// Would load and render actual templates
	var sb strings.Builder
	sb.WriteString(fmt.Sprintf("Template: %s\n", name))
	for k, v := range data {
		sb.WriteString(fmt.Sprintf("%s: %v\n", k, v))
	}
	return sb.String()
}

func (s *EmailService) generateVerificationToken(user *models.User) string {
	// Would generate actual secure token
	return fmt.Sprintf("verify_%s", user.ID.String())
}

type BulkEmailResult struct {
	Sent   int
	Failed int
	Errors []BulkEmailError
}

type BulkEmailError struct {
	Email string
	Error string
}

// Template parsing helper
func parseTemplate(name, content string) (*template.Template, error) {
	return template.New(name).Parse(content)
}

File: tests/fixtures/python-django/myapp/services/email_service.py
# myapp/services/email_service.py - Email service

from typing import Optional, List, Dict, Any
from django.conf import settings
from django.core.mail import send_mail, EmailMultiAlternatives
from django.template.loader import render_to_string
from django.utils.html import strip_tags
import logging

logger = logging.getLogger(__name__)


class EmailService:
    """Service for sending emails."""

    def __init__(self):
        self.from_email = getattr(settings, 'DEFAULT_FROM_EMAIL', 'noreply@example.com')
        self.site_name = getattr(settings, 'SITE_NAME', 'MyApp')
        self.base_url = getattr(settings, 'BASE_URL', 'http://localhost:8000')

    def send_verification_email(self, user) -> bool:
        """Send email verification link to user."""
        token = self._generate_verification_token(user)
        verification_url = f"{self.base_url}/verify-email?token={token}"

        context = {
            'user': user,
            'verification_url': verification_url,
            'site_name': self.site_name,
        }

        return self._send_template_email(
            to_email=user.email,
            subject=f'Verify your email for {self.site_name}',
            template='emails/verify_email.html',
            context=context
        )

    def send_password_reset_email(self, user, token: str) -> bool:
        """Send password reset link to user."""
        reset_url = f"{self.base_url}/reset-password?token={token}"

        context = {
            'user': user,
            'reset_url': reset_url,
            'site_name': self.site_name,
            'expires_in': '24 hours',
        }

        return self._send_template_email(
            to_email=user.email,
            subject=f'Reset your password for {self.site_name}',
            template='emails/password_reset.html',
            context=context
        )

    def send_welcome_email(self, user) -> bool:
        """Send welcome email to new user."""
        context = {
            'user': user,
            'site_name': self.site_name,
            'login_url': f"{self.base_url}/login",
        }

        return self._send_template_email(
            to_email=user.email,
            subject=f'Welcome to {self.site_name}!',
            template='emails/welcome.html',
            context=context
        )

    def send_suspension_notice(
        self,
        user,
        reason: str,
        duration_days: Optional[int] = None
    ) -> bool:
        """Send account suspension notification."""
        context = {
            'user': user,
            'reason': reason,
            'duration_days': duration_days,
            'site_name': self.site_name,
            'support_email': getattr(settings, 'SUPPORT_EMAIL', 'support@example.com'),
        }

        return self._send_template_email(
            to_email=user.email,
            subject=f'Your {self.site_name} account has been suspended',
            template='emails/suspension_notice.html',
            context=context
        )

    # Q: What's the best approach for handling email delivery failures and retries?
    def send_bulk_email(
        self,
        users: List,
        subject: str,
        template: str,
        extra_context: Optional[Dict[str, Any]] = None
    ) -> Dict[str, int]:
        """Send email to multiple users."""
        results = {'sent': 0, 'failed': 0}

        for user in users:
            context = {
                'user': user,
                'site_name': self.site_name,
                **(extra_context or {})
            }

            success = self._send_template_email(
                to_email=user.email,
                subject=subject,
                template=template,
                context=context
            )

            if success:
                results['sent'] += 1
            else:
                results['failed'] += 1

        return results

    def send_notification_email(
        self,
        user,
        notification_type: str,
        data: Dict[str, Any]
    ) -> bool:
        """Send notification email based on type."""
        templates = {
            'new_login': 'emails/new_login.html',
            'password_changed': 'emails/password_changed.html',
            'profile_updated': 'emails/profile_updated.html',
            'security_alert': 'emails/security_alert.html',
        }

        template = templates.get(notification_type)
        if not template:
            logger.warning(f'Unknown notification type: {notification_type}')
            return False

        subjects = {
            'new_login': f'New login to your {self.site_name} account',
            'password_changed': f'Your {self.site_name} password was changed',
            'profile_updated': f'Your {self.site_name} profile was updated',
            'security_alert': f'Security alert for your {self.site_name} account',
        }

        context = {
            'user': user,
            'site_name': self.site_name,
            **data
        }

        return self._send_template_email(
            to_email=user.email,
            subject=subjects[notification_type],
            template=template,
            context=context
        )

    def _send_template_email(
        self,
        to_email: str,
        subject: str,
        template: str,
        context: Dict[str, Any]
    ) -> bool:
        """Send email using template."""
        try:
            html_content = render_to_string(template, context)
            text_content = strip_tags(html_content)

            email = EmailMultiAlternatives(
                subject=subject,
                body=text_content,
                from_email=self.from_email,
                to=[to_email]
            )
            email.attach_alternative(html_content, 'text/html')
            email.send()

            logger.info(f'Email sent successfully to {to_email}')
            return True

        except Exception as e:
            logger.error(f'Failed to send email to {to_email}: {str(e)}')
            return False

    def _generate_verification_token(self, user) -> str:
        """Generate email verification token."""
        from django.contrib.auth.tokens import default_token_generator
        return default_token_generator.make_token(user)

    def _send_simple_email(
        self,
        to_email: str,
        subject: str,
        message: str
    ) -> bool:
        """Send simple text email."""
        try:
            send_mail(
                subject=subject,
                message=message,
                from_email=self.from_email,
                recipient_list=[to_email],
                fail_silently=False
            )
            return True
        except Exception as e:
            logger.error(f'Failed to send simple email: {str(e)}')
            return False
